### PR TITLE
Feat: Add subnet to a existing virtual network

### LINKED ISSUE #61: [AVM Module Issue]: Usage of a var instead of a local
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Module specific issue

- [X] I confirm that this issue is about this AVM module and not about the resource itself, or how to use Terraform.

### Issue Type?

I'm not sure

### (Optional) Module Version

0.1.4

### (Optional) Correlation Id

_No response_

### Description

main.telemetry.tf uses the local enable_telemetry on line 2 and then uses the var with same name on line 10. This seems like a typo (something forgotten) and it is confusing. Is that the desired state or should it really be a reference to the local on line 10 as well?

### PR DESCRIPTION
**Pull request by herms14**
Fixed telemetry variable and added feature to attach a subnet to a existing Virtual network

## Type of Change

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `locals.version.tf.json`:
    - [X] Someone has opened a bug report issue. Close #61 and #43 
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `locals.version.tf.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `locals.version.tf.json`.
  - [X] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] I did run all  [pre-commit](https://azure.github.io/Azure-Verified-Modules/contributing/terraform/terraform-contribution-flow/#5-run-pre-commit-checks) checks

### CODE REVIEW

File: examples/existing_vnet/.terraform-docs.yml
Comment (matt-FFFFFF) [inline examples/existing_vnet/.terraform-docs.yml:1]: This file should not be here. The terraform docs files should be in the root and in the examples dir.

File: examples/existing_vnet/Makefile
Comment (matt-FFFFFF) [inline examples/existing_vnet/Makefile:1]: This file should not be here

File: variables.tf [hunk 1]
```diff
@@ -57,6 +57,19 @@ If it is set to false, then no telemetry will be collected.
 DESCRIPTION
 }
 
+variable "existing_virtual_network" {
```
Comment (matt-FFFFFF) [inline variables.tf:60]: Please can you add _resource_id on the end?


This makes it super clear what we are looking for.

File: main.tf [hunk 2]
```diff
@@ -73,7 +75,7 @@ resource "azurerm_monitor_diagnostic_setting" "example" {
   }
 
   name                           = each.value.name != null ? each.value.name : "defaultDiagnosticSetting"
-  target_resource_id             = azurerm_virtual_network.vnet.id
+  target_resource_id             = var.existing_virtual_network != null ? var.existing_virtual_network.id : azurerm_virtual_network.vnet[0].id
```
Comment (matt-FFFFFF) [inline main.tf:78]: We can use coalesce() here I think...

I would move this into a local so that you can use in multiple places

File: main.tf [hunk 3]
```diff
@@ -47,15 +49,15 @@ resource "azurerm_management_lock" "this" {
...
+  scope      = azurerm_virtual_network.vnet[count.index].id
 }
 
 # Assigning Roles to the Virtual Network based on the provided configurations.
 resource "azurerm_role_assignment" "vnet_level" {
   for_each = var.role_assignments
 
   principal_id                           = each.value.principal_id
-  scope                                  = azurerm_virtual_network.vnet.id
+  scope                                  = var.existing_virtual_network != null ? var.existing_virtual_network.id : azurerm_virtual_network.vnet[0].id
```
Comment (matt-FFFFFF) [inline main.tf:60]: See below - move to local and consider using coalesce

File: main.tf [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Azure Generic vNet Module
-# Creating a Virtual Network with the specified configurations.
+# Conditionally create a Virtual Network with the specified configurations.
```
Comment (matt-FFFFFF) [inline main.tf:2]: Add a moved block to tell TF that azurerm_virtual_network.vnet has moved to azurerm_virtual_network.vnet[0]

**Comment by matt-FFFFFF** [comment]
Great work!

A few comments for consideration.

- adding a moved block
- adding more locals to determine the vnet details. These locals would provide the values for either the existing or the new vnet, depending on whether the existing input is null.

Coalesce is a good function to use here!

File: examples/existing_vnet/.terraform-docs.yml
Comment (kewalaka) [inline examples/existing_vnet/.terraform-docs.yml:1]: the examples don't need their own terraform docs file - it is made by calling ``./avm docs`` from the root of the project - this creates the docs for each example.

File: examples/existing_vnet/Makefile
Comment (kewalaka) [inline examples/existing_vnet/Makefile:1]: again this is not necessary - the ``avm`` file in the root of the project covers makefile requirements

File: examples/existing_vnet/_footer.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+## Usage
```
Comment (kewalaka) [inline examples/existing_vnet/_footer.md:1]: usage should be the header not the footer

File: examples/existing_vnet/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,9 @@
+output "subnets" {
+  description = "Information about the subnets created in the module."
+  value       = module.create_subnet.subnets
+}
+
+output "vnet_id" {
```
Comment (kewalaka) [inline examples/existing_vnet/outputs.tf:6]: this should be ``id`` rather than ``vnet_id``

File: examples/existing_vnet/providers.tf [hunk 7]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  required_version = ">= 1.5.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.7.0, < 4.0.0"
```
Comment (kewalaka) [inline examples/existing_vnet/providers.tf:6]: ``version = "~> 3.7.0"`` is a simpler way to express the same thing (any version up to but not including the next major version).

File: examples/existing_vnet/variables.tf [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#Specifies the location of the resource group.
+variable "rg_location" {
```
Comment (kewalaka) [inline examples/existing_vnet/variables.tf:2]: ``location`` is the standard variable name for specifying locations.  Resources should be created in the same location as the parent resource group.

File: examples/existing_vnet/variables.tf [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  type        = string
+  default     = "eastus"
+  description = <<DESCRIPTION
+This variable defines the Azure region where the resource group will be created.
+The default value is "westus".
+DESCRIPTION
+}
+
+#Specifies the location of the virtual network.
+variable "vnet_location" {
```
Comment (kewalaka) [inline examples/existing_vnet/variables.tf:12]: this is redundant imo - just have a single ``location`` variable.   Resources should be created in the same location as the parent resource group.

File: outputs.tf [hunk 10]
```diff
@@ -13,7 +13,7 @@ output "subnets" {
 
 output "virtual_network_id" {
```
Comment (kewalaka) [inline outputs.tf:14]: should be called ``id`` to be AVM-aligned.  Note - breaking change!

File: outputs.tf [hunk 11]
```diff
@@ -13,7 +13,7 @@ output "subnets" {
...
   description = "The resource ID of the virtual network."
-  value       = azurerm_virtual_network.vnet.id
+  value       = var.existing_virtual_network != null ? var.existing_virtual_network.id : azurerm_virtual_network.vnet[0].id
 }
 
 output "vnet_resource" {
```
Comment (kewalaka) [inline outputs.tf:19]: should be called ``resource`` to be AVM aligned.  Note - breaking change!

File: examples/existing_vnet/providers.tf (hunk 7 above)
Comment (matt-FFFFFF) [reply examples/existing_vnet/providers.tf:6]: Rather `~> 3.7`

**Comment by herms14** [comment]
thanks for the feedback guys. been busy with other stuffs lately so will be able to woro on these change probably by next week

**Comment by kewalaka** [comment]
@herms14 i think found when i was doing some testing that if you create multiple subnets it re-creates them with each plan.  I think there is a conflict within the module that forces this.  I haven't had chance to test it yet & was planning to raise an issue if so, but just wanted to give you the heads up, as obviously re-creating subnets for each plan is pretty serious.

if you've got chance to take a look that would be great - if you can't spot the issue i'll see if i can re-pro.

**Comment by kewalaka** [comment]
i've got a potentially better way to do this, which also address #71 

currently testing it here; https://github.com/kewalaka/terraform-azurerm-avm-res-network-virtualnetwork/tree/feat/use-azapi-for-subnets

**Comment by kewalaka** [comment]
#74 supersedes this and addresses some other issues as noted in that PR.

**Comment by matt-FFFFFF** [comment]
Closing in favour of #74

**Pull request closed without merging**